### ISSUE TITLE
Monitor callback never fired

### ISSUE THREAD

**Issue by luis4glfw** [glfw/glfw#784]
The monitor callback set with glfwSetMonitorCallback() is never fired on my system.

I have a nVidia GTX970 with 353.06 drivers on Win 7 x64.

I'm using the latest stable version of glfw, 3.2.

```
3.2.0 Win32 WGL EGL MinGW DLL
NVIDIA Corporation
GeForce GTX 970/PCIe/SSE2
4.5.0 NVIDIA 353.06
```

Running the test program events.c, I tried to enable/disable a second monitor from the nVidia control panel, the monitor is attached/removed as expected to/from the OS but no event is fired in the events test program.

I also tried to physically connect/disconnect the monitor from the graphic card, still the same.

I also tried to manually call glfwGetMonitors() from my own code (by pressing a function key) after I attached/detached the second monitor, but the list of reported monitors is not updated and it's the one detected at the moment the program was launched.

I have to exit the program and relaunch it to make glfwGetMonitors() aware of the changed configuration.

Also I tried in a simple API program with a single window, to see what messages were sent to the window.
In winproc I'm testing for WM_DEVICECHANGE and WM_DISPLAYCHANGE

I get a WM_DEVICECHANGE when I connect/disconnect a USB joystick
I get a WM_DISPLAYCHANGE when I add/remove a monitor in the nVidia control panel, but no WM_DEVICECHANGE

In my case, to make all work I had to change windowProc() in win32_windows.c adding WM_DISPLAYCHANGE to the second switch(), the one associated to the MAIN window (not the helper one).

```
switch (uMsg)
    {
    ...
    }

switch (uMsg)
    {
        // start
        case WM_DISPLAYCHANGE:
        {        
            _glfwInputMonitorChange();
            return 0;
        }
        // end

        case WM_SETFOCUS:
       ...
       etc
       ...
```

This way the callback work for bot attacth and detach and the monitor list has been refreshed when I call glfwGetMonitors()

User @dougbinks reports 

> I've just tested this and on my Windows 8.1 system with Radeon drivers and latest GLFW 3.2 from Github I get the same as you.
> Investigating further it appears the WM_DEVICECHANGE message is being sent with a valid window on my system, so copying the code for device change (testing DBT_DEVNODES_CHANGED only) outside the if(!window) and in the second switch works for me.

So his modification is different:

```
switch (uMsg)
    {
    ...
    }

switch (uMsg)
    {       
        // start
        case WM_DEVICECHANGE:
        {
            if (wParam == DBT_DEVNODES_CHANGED)
            {
                _glfwInputMonitorChange();
                return TRUE;
            }
        }
        // end

        case WM_SETFOCUS:           
       ...
       etc
       ...
```

The above works for him, but NOT for me.

The original thread in the forum is here -> http://discourse.glfw.org/t/monitor-callback-never-fired/678/1

Thank you

**Comment by elmindreda** [comment]
Thank you for the detailed bug report! Will look into this for 3.3. If it really can't be helped then I'll disable monitor object caching.

**Comment by elmindreda** [comment]
This should be fixed now with 97dbd8b63bbd15bb781e54a670b183dbd59f1bf0, for both physical and software-based connection and disconnection.
